### PR TITLE
[_534] implement replica truncate

### PR DESCRIPTION
**Pull request by d-w-moore**
still needs a test

### CODE REVIEW

File: irods/manager/data_object_manager.py
Comment (korydraughn) [inline irods/manager/data_object_manager.py:463]: Are there any conventions around naming functions?

I've noticed that some function names are closer to their C counterparts and others not so much.

So, why call this `replica_truncate` vs `truncate`?

Comment (d-w-moore) [reply irods/manager/data_object_manager.py:463]: there's already a data_objects.truncate( ) , yes, although it's based on the existing (but deprecated I think?) truncate API

Comment (korydraughn) [reply irods/manager/data_object_manager.py:463]: Ah ok. Is the `truncate` function in PRC marked as deprecated? If not, please capture that in an issue.

As for this conversation, I think we're good. Resolving.

Comment (d-w-moore) [reply irods/manager/data_object_manager.py:463]: Will do, the deprecation could happen in PRC v2.1.1  I guess but with a  3.0.0 milestone for implementing that goal, because PRC will still have to work with older servers and applications.

Comment (korydraughn) [reply irods/manager/data_object_manager.py:463]: Sounds good to me.

Comment (trel) [reply irods/manager/data_object_manager.py:463]: Why not 2.1.0?

Comment (korydraughn) [reply irods/manager/data_object_manager.py:463]: Either is fine with me. Deprecate in 2.x, remove in 3.x.

Comment (d-w-moore) [reply irods/manager/data_object_manager.py:463]: > Why not 2.1.0?

Sure.  `2.1.0` still has a few PR's to go, to be complete, and deprecating there would be fine too assuming we're in no hurry.

File: irods/manager/data_object_manager.py
Comment (alanking) [inline irods/manager/data_object_manager.py:540]: This isn't *strictly speaking* true because replica_truncate exists and is usable in 4.3.2. However, https://github.com/irods/irods/issues/7918 describes why it's not as useful as it could be (the output string is empty on success even though it's supposed to contain information).

So, we should either keep this as it is with an explanatory comment to the effect of the one I left above, or we should change this to 4.3.2 and handle its attendant issues (if possible). I don't recall whether there was an issue with using the 4.3.2 endpoint here, but if it's enough of a problem to make it unusable, then leaving this as 4.3.3 with an explanatory comment is fine.

Just my two cents

**Review by alanking — COMMENTED** [review]
Looking good so far

**Comment by alanking** [comment]
@d-w-moore - Please mark as ready for review when ready for review.

**Comment by d-w-moore** [comment]
Ok - this is now ready for review.

File: irods/exception.py
Comment (korydraughn) [inline irods/exception.py:72]: Why change the base type?

File: irods/test/data_obj_test.py
Comment (korydraughn) [inline irods/test/data_obj_test.py:2365]: Consider adding a test that proves increasing the size of a replica works as intended.

**Review by korydraughn — COMMENTED** [review]
Looks good overall.

File: irods/test/data_obj_test.py
Comment (alanking) [reply irods/test/data_obj_test.py:2365]: I would add that we should get one test with an error response to see what that looks like as well (e.g. a negative size or specifying a resource which does not have a replica of the target data object). Just so we exercise the exceptions.

Comment (korydraughn) [reply irods/test/data_obj_test.py:2365]: Agreed.

File: irods/exception.py
Comment (d-w-moore) [reply irods/exception.py:72]: A number of Python iRODS Client exceptions are derived from it, in particular `OperationNotSupported`. My own theory is that the PRC was originally supposed to support  a substitute for, or customization of, iCommands, and that a PycommandsException meant you were attempting something that was unsupported.  I'd be equally happy reverting the `NotImplementedInIRODSServer` base class back to `RuntimeError`, changing it to `Exception`, or giving it the same base class as `NotImplementedInIRODSServer` (which was the motivation for this change).

@alanking and i discussed this earlier.  the advantage to this approach might be that an application developer could use `PycommandsException` as a catch-all for both `NotImplementedInIRODSServer` and `OperationNotSupported`.

Comment (korydraughn) [reply irods/exception.py:72]: Ok. Can this break existing code?

Comment (alanking) [reply irods/exception.py:72]: `NotImplementedInIRODSServer` was introduced in commit https://github.com/irods/python-irodsclient/commit/3015e40a1c1c8868ec22efe35751e3d0d85bebd8 which is not in a release.

Comment (d-w-moore) [reply irods/exception.py:72]: Right, otherwise I wouldn't have considered this change without a lot more todo.  (Some doc changes at least)

Comment (korydraughn) [reply irods/exception.py:72]: Cool. Carry on.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:2365]: We get SYS_REPLICA_INACCESSIBLE (168000) if we target a resource not housing a replica of the data obj in question.  That seems good to me. Shall we have a test asserting that particular type of exception / error?

Comment (alanking) [reply irods/test/data_obj_test.py:2365]: Sounds good to me. The important part is that we are exercising the error mechanism of the PRC method, but we can /should also assert that the error code is correct, yes.

Comment (d-w-moore) [reply irods/test/data_obj_test.py:2365]: And `UNIX_FILE_TRUNCATE_ERR` with OS errno 22 when a negative size is fed in ....

File: irods/manager/data_object_manager.py
Comment (d-w-moore) [reply irods/manager/data_object_manager.py]: This seems resolved.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:2400]: Done! Wrote test including a reasonable range of errors one might see appear in this API / endpoint....

File: irods/manager/data_object_manager.py
Comment (d-w-moore) [reply irods/manager/data_object_manager.py]: Corrected.

**Comment by d-w-moore** [comment]
Squashed - and all concerns so far seemingly addressed - @alanking @korydraughn please confirm by reading new test that remaining conversation is resolved by this addition.

File: irods/test/data_obj_test.py
Comment (alanking) [inline irods/test/data_obj_test.py:2391]: Are you able to capture the output when these exceptions are thrown, or do we just get the error code?

If it's possible to get the JSON output in the error case, let's try to capture it and... well, we don't need to assert the contents per se, but we can make sure that it's valid JSON, perhaps. Thoughts?

File: irods/test/data_obj_test.py
Comment (alanking) [inline irods/test/data_obj_test.py:2419]: Can we capture the JSON output here and perhaps assert that the replica number and resource hierarchy are as expected?

**Review by alanking — COMMENTED** [review]
Good. Had a couple of suggestions. This is close

File: irods/manager/data_object_manager.py
Comment (korydraughn) [reply irods/manager/data_object_manager.py:540]: Let's keep github issue numbers out of the error messages.

The additional wording sounds good to me.

I wonder if it would be better to state that the method isn't supported first and then give the reason. I don't have strong feelings on that. Keeping the suggestion or swapping the order accomplish the job.

Comment (korydraughn) [reply irods/manager/data_object_manager.py:540]: Yep. That's it.

I defer to you and @d-w-moore.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:2391]: Tried stepping through, and it appears the response is parsed as an error, rather than  as a routine response with JSON output expected.  SHould the PRC be looking for the JSON output in addition to an error code?

Comment (alanking) [reply irods/test/data_obj_test.py:2391]: Gotcha. The JSON output in error cases would provide an error message from the server that could be useful for applications built on top of this one. I guess PRC is more relying on the rError stack for things like that, so it can't be helped. Let's just let it ride because it's fine as is. Resolving.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:2419]: It appears the "boilerplate" code typically used to return a response does not return any embedded JSON.  It might be accessible in some future version of the PRC, if processing / testing of such return messages is  useful in the client.

Comment (alanking) [reply irods/test/data_obj_test.py:2419]: So... I guess I'm confused, then. What does this line in `data_object_manager.replica_truncate` do (line 566):
```python
        return json.loads(msg.myStr)
```
It looks to me like this method returns a JSON object.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:2391]: Ok.  Just wondering, then, will the custom of using an rError stack soon be in disfavor, with the useful information coming back via the JSON message response instead?  Is that perhaps the "new way"?

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:2419]: I was looking at the wrong code, sorry! You are correct.

File: irods/test/data_obj_test.py
Comment (alanking) [reply irods/test/data_obj_test.py:2391]: > Ok. Just wondering, then, will the custom of using an rError stack soon be in disfavor, with the useful information coming back via the JSON message response instead? Is that perhaps the "new way"?

I redirect this question to @korydraughn. Many of the newer APIs are using JSON output, but not all. I don't think we've settled on a direction yet. It does seem (to me) that rError usage has fallen out of favor.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py]: Ok we're doing that now.  Should I also assert msg is empty for the routine ( non-error )case?

File: irods/test/data_obj_test.py
Comment (korydraughn) [reply irods/test/data_obj_test.py:2391]: I wouldn't say one way is favored over the other. It depends on the design of the API.

iRODS provides multiple ways of doing things. For the replica truncate API, returning JSON via an out parameter made sense. A different API may pass things back via the rError stack and there's nothing wrong with that _(even if it's not preferred)_.

The important thing is that we continue to document expected usage and behavior so users don't get confused and lose confidence in the interface.

Comment (d-w-moore) [reply irods/test/data_obj_test.py:2391]: Ah, ok.  I may need to bake into the client some way of getting to the JSON out parameter in the case of error.  Presently, the irods.message interface throws an exception rather than allowing application access to the return message that causes it.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py]: Check. ok, resolving

File: irods/test/data_obj_test.py
Comment (alanking) [reply irods/test/data_obj_test.py:2391]: Should we push this off for a later release, or do it now? Asking because I resolved the conversation but it looks like we're looking into more work... Can unresolve if that seems helpful.

Comment (korydraughn) [reply irods/test/data_obj_test.py:2391]: Let's punt on replica truncate and get the PRC released.

The 2.1.0 milestone has 25 closed issues which is plenty.

Comment (korydraughn) [reply irods/test/data_obj_test.py:2391]: After some discussion, we've decided to open an issue for 3.0 to investigate allowing the out parameter to be captured on error and allow this PR to be merged.

Comment (alanking) [reply irods/test/data_obj_test.py:2391]: The new issue is #606. There is a PR to add the functionality in #607. I think this can be resolved.

**Review by alanking — COMMENTED** [review]
Pending resolution of open conversation... I think we can go ahead and squash this (no #s).

**Comment by korydraughn** [comment]
Agreed.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:2391]: Yes, agree, ... resolving.

**Comment by d-w-moore** [comment]
Note to self (@d-w-moore ) write another test in #607 for replica_truncate & possibly other JSON oriented APIs. Pending merge of that PR for [milestone v2.1.1](https://github.com/irods/python-irodsclient/milestone/5)

**Comment by korydraughn** [comment]
If the tests are passing, squash to taste. No pounds.

**Review by alanking — APPROVED** [review]
Seems good to me. I say # it. Or, wait for other approval(s)

**Comment by d-w-moore** [comment]
Will wait for other approvals, running final test.

**Review by korydraughn — APPROVED** [review]
Pound it if final testing passes.

**Comment by d-w-moore** [comment]
> Pound it if final testing passes.

Tests pass - now pounded.

**Pull request merged**
